include/cilantro/core/covariance.hpp

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include <algorithm>
+#include <limits>
+
+#include <cilantro/core/data_containers.hpp>
+#include <cilantro/core/nearest_neighbors.hpp>
+#include <cilantro/utilities/random.hpp>
+
+namespace cilantro {
+
+    template <typename ScalarT, ptrdiff_t EigenDim>
+    class Covariance {
+    public:
+        EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+        Covariance() = default;
+
+        inline bool operator()(const ConstVectorSetMatrixMap<ScalarT,EigenDim> &points, Vector<ScalarT,EigenDim>& mean, Eigen::Matrix<ScalarT,EigenDim,EigenDim>& cov) const {
+            if (points.cols() < points.rows()) return false;
+            mean = points.rowwise().mean();
+            auto centered = points.colwise() - mean;  // Lazy evaluation
+            cov.noalias() =  (ScalarT)(1.0)/(points.cols() - 1) * (centered * centered.transpose());
+            return true;
+        }
+
+        template <typename NeighborhoodResultIteratorT>
+        inline bool operator()(const ConstVectorSetMatrixMap<ScalarT,EigenDim> &points, NeighborhoodResultIteratorT begin, NeighborhoodResultIteratorT end, Vector<ScalarT,EigenDim>& mean, Eigen::Matrix<ScalarT,EigenDim,EigenDim>& cov) const {
+            size_t size = std::distance(begin, end);
+            if (size < points.rows()) return false;
+
+            mean.setZero();
+            for (NeighborhoodResultIteratorT it = begin; it != end; ++it) {
```

Comment:
Ideally, we should optionally parallelize the for loops in this file.

Reply:
Hi Wannes, I added optional parallelization in aab6719ce66da89a482ca2d13f02ca07a3e9989d (and refactored PCA to make use of `Covariance`). I also slightly changed the `MinimumCovarianceDeterminant` API in that now `operator()` does not alter the Neighborhood input (makes a copy internally for Mahalanobis computation and sorting). Performance hit was <1% for the robust normal estimation example. If that breaks anything, please let me know :D

Reply:
Awesome! Thanks!